### PR TITLE
[IMP] website: convert theme_preview_kanban, theme_preview_form, iframe widget to owl

### PR DESCRIPTION
**Pull request by younn-o**
This PR adapts the following fields and views to OWL:
- theme_preview_form, 
- theme_preview_kanban, 
- website_publish_button, 
- website_cookiesbar_field, 
- FieldIframePreview.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by detrouxdev** [comment]
@qsm-odoo I've adapted everything asked to OWL. 

Unfortunately I wasn't able to remove the legacy `website_redirect_button` as there's still at least one view that needs it. (Invoice -> Create a new Invoice -> Select a partner -> Expand Partner view -> Legacy view with a website redirect button if website_partner is installed.) 

I marked them as TODO so that we can easily remove them once that view is converted to OWL.

File: addons/website/static/src/components/views/theme_preview_form.js [hunk 1]
```diff
@@ -1,157 +1,109 @@
...
-        const $switchModeButton = $(qweb.render('website.ThemePreview.SwitchModeButton'));
-        this.controlPanelProps.cp_content.$pager = $switchModeButton;
-        return props;
-    },
+    install() {
+        return this.env.onClickViewButton({
+            clickParams: {
+                            name: "button_choose_theme",
+                            type: "object",
+                        },
```
Comment (qsm-odoo) [inline addons/website/static/src/components/views/theme_preview_form.js:71]: Unindent this to align with clickParams:

```javascript
clickParams: {
    name: "button_choose_theme"
    type: "object",
},
```
Although, can you explain to me a bit more this `onClickViewButton` stuff? I am not sure to understand the explanations.

File: addons/website/views/website_views.xml [hunk 2]
```diff
@@ -555,12 +555,10 @@
             <field name="model">ir.module.module</field>
             <field name="mode">primary</field>
             <field name="arch" type="xml">
-                <form create="false" edit="false" delete="0" js_class="theme_preview_form">
+                <form create="false" edit="false" delete="0" js_class="theme_preview_form" class="o_view_form_theme_preview">
```
Comment (qsm-odoo) [inline addons/website/views/website_views.xml:558]: Should the class be added by the JS, seems strange to allow `js_class="theme_preview_form"` without the class no?

File: addons/website/static/src/xml/website_widget.xml [hunk 3]
```diff
@@ -2,15 +2,19 @@
...
+            <iframe
+                frameBorder="0"
+                class="d-block"
+                height="100%"
+                width="100%"
+                t-att-src="props.value"/>
+            <img alt="phone"
+                 class="img_mobile"
+                 style="display: none;"
+                 src="/website/static/src/img/phone.png"/>
```
Comment (qsm-odoo) [inline addons/website/static/src/xml/website_widget.xml:16]: Is this needed to have it in display: none and show it via the class? Can't this just be not rendered at all? Maybe it flickers if we do that?

File: addons/website/views/website_views.xml [hunk 4]
```diff
@@ -555,12 +555,10 @@
...
                   <div class="o_preview_frame h-100">
-                        <field name='url' widget='iframe'/>
-                        <img alt='phone' class='img_mobile' style='display:none' src="/website/static/src/img/phone.png"/>
+                        <field name='url' widget='iframe' class="d-block m-0 h-100"/>
```
Comment (qsm-odoo) [inline addons/website/views/website_views.xml:560]: While you are touching this line, take the opportunity to change the `'` into `"` :+1:

File: addons/website/static/src/components/fields/publish_button.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    }
+    get hoverClass() {
+        return 'o_stat_text o_hover ' + (this.props.value ? 'text-danger' : 'text-success');
+    }
+    get value() {
+        return this.props.value ? this.env._t("Published") : this.env._t("Unpublished");
+    }
+    get valueClass() {
+        return 'o_stat_text o_not_hover ' + (this.props.value ? 'text-success' : 'text-danger');
+    }
```
Comment (qsm-odoo) [inline addons/website/static/src/components/fields/publish_button.js:19]: Is any of this necessary to do in JS? Why is this not in the XML template directly?

File: addons/website/static/src/components/fields/redirect_field.xml [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
-        <button type="button" class="oe_stat_button btn btn-light w-100 mb-0 py-0 border-0 border-start border-bottom rounded-0 text-start text-nowrap text-capitalize"
-            t-on-click="onClick"
-            t-att-aria-label="info"
-            t-att-title="info">
-            <i class="fa fa-fw o_button_icon fa-globe" t-att-class="this.props.value ? 'text-success' : 'text-danger'"/>
-            <div class="o_stat_info">
-                <span class="o_stat_text">Go to<br/>Website</span>
-            </div>
-        </button>
-    </t>
```
Comment (qsm-odoo) [inline addons/website/static/src/components/fields/redirect_field.xml]: Mmh, I would be tempted to let you do that (I am a strong partisan of letting people break the history if it means using code conventions or improving readability) but I would prefer to do this only in dedicated commits (so that when we do use the git history we directly know if we are going to hit a commit that is just moving code around).

File: addons/website/static/src/components/fields/res_config_settings.xml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates id="template" xml:space="preserve">
+<t t-name="website.res_config_settings.CookieDialog" owl="1">
```
Comment (qsm-odoo) [inline addons/website/static/src/components/fields/res_config_settings.xml:3]: Add blank line after <templates (around each template)

File: addons/website/static/src/js/backend/button.js [hunk 8]
```diff
@@ -7,46 +7,7 @@ var field_registry = require('web.field_registry');
...
-        var hover = this.value ? _t("Unpublish") : _t("Publish");
-        var valColor = this.value ? 'text-success' : 'text-danger';
-        var hoverColor = this.value ? 'text-danger' : 'text-success';
-        var $val = $('<span>').addClass('o_stat_text o_not_hover ' + valColor).text(text);
-        var $hover = $('<span>').addClass('o_stat_text o_hover ' + hoverColor).text(hover);
-        this.$el.append($val).append($hover);
-    },
-});
-
+// TODO: Remove me when Partner view is adapted to OWL
```
Comment (qsm-odoo) [inline addons/website/static/src/js/backend/button.js:10]: > Unfortunately I wasn't able to remove the legacy website_redirect_button as there's still at least one view that needs it. (Invoice -> Create a new Invoice -> Select a partner -> Expand Partner view -> Legacy view with a website redirect button if website_partner is installed.)

Your comment says "remove me", I guess it's more "adapt me", right? Maybe you can already do the adaptation in the same file and keep the old and rename it as `website_redirect_button_legacy` (and use that one in problematic non converted view?) Maybe even move its whole code in the problematic app if it's only one place?

File: addons/website/static/src/xml/website.res_config_settings.xml [hunk 9]
```diff
@@ -1,18 +0,0 @@
...
-            consent in your country.
-        </p>
-        <p>
-            For session cookies, authentification and analytics,
-            <b>you do not need to ask for the consent</b> (see e.g. Opinion
-            04/2012 on Cookie Consent Exemption by the EU Art.29 WP).
-        </p>
-    </main>
-</t>
-</templates>
```
Comment (qsm-odoo) [inline addons/website/static/src/xml/website.res_config_settings.xml:18]: Should this file first be moved in the dedicated move commit to preserve some history ? :+1:

File: addons/website/static/tests/website_tests.js [hunk 10]
```diff
@@ -6,6 +6,7 @@ var testUtils = require("web.test_utils");
 
 var createView = testUtils.createView;
 
+// TODO: remove me when legacy redirect button is removed
```
Comment (qsm-odoo) [inline addons/website/static/tests/website_tests.js:9]: Strange, we already had a "redirect_field_tests.js" file alongside a "website_tests.js" file testing only redirect buttons? :laughing: 

Note: even more reason to do the thing I said and have both website_redirect_button and website_redirect_button_ledacy, I am sure we will forget about this file otherwise ^^

File: addons/website/static/src/js/backend/button.js [hunk 11]
```diff
@@ -111,6 +72,5 @@ var WidgetWebsiteButtonIcon = AbstractField.extend({
 });
 
 field_registry
-    .add('website_redirect_button', WidgetWebsiteButtonIcon)
-    .add('website_publish_button', WebsitePublishButton);
+    .add('website_redirect_button', WidgetWebsiteButtonIcon);
```
Comment (qsm-odoo) [inline addons/website/static/src/js/backend/button.js:75]: Commit message: "This commit adapts the website_redirect_button and website_published_button to OWL."

Unless I totally misunderstand this whole commit, this does not yet convert `website_redirect_button` (but will if you do `website_redirect_button` and `website_redirect_button_legacy` as explained :+1:)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Cool diff, some questions and remarks then I'll test and merge :+1:

File: addons/website/static/src/components/views/theme_preview_form.js (hunk 1 above)
Comment (detrouxdev) [reply addons/website/static/src/components/views/theme_preview_form.js]: I figured out how to make it a ViewButton component. So this code was dropped, but in summary, it was to use the same logic of clicking on a button from a record.

File: addons/website/static/src/js/backend/button.js (hunk 11 above)
Comment (detrouxdev) [reply addons/website/static/src/js/backend/button.js:75]: The commit message was wrong however, there is already a `website_redirect_button` that was implemented here https://github.com/odoo/odoo/commit/9434969d5dc06aadcdbbfc600514a0bdb5635ea2 .

(Hence the tests already existing) I assume they kept the old one because of the view. 

For moving it to the module that still uses it, I'm not so certain which one still does. The way I found about the use of the legacy button was through a test. I removed the button and a test failed. That test was in the accounting module when editing a partner in a modal view.

I'm not confident enough to say that only these 2 modules use it, as a few views still need to be converted, and I'm not certain I know of all the flows to access the view that uses a `website_redirect_button`.

For the publish widget I feel confident that we can remove it as it's not as widely used.

I don't mind moving the legacy button in a separate `legacy` folder on website though if that's OK with you ?

Comment (qsm-odoo) [reply addons/website/static/src/js/backend/button.js:75]: > I'm not confident enough to say that only these 2 modules use it, as a few views still need to be converted, and I'm not certain I know of all the flows to access the view that uses a website_redirect_button.

Well somebody will need to check all that at some point :wink: What's stopping you to check it ? :thinking: Is not this just about checking all uses in the code one by one (without having to know the flows) and if they are not in a custom view (js_class ?), then they can use the new one, otherwise they should use the old one? It would be useful to know as maybe the teams related to the old implementation may never care otherwise (because maybe they don't know).

Comment (detrouxdev) [reply addons/website/static/src/js/backend/button.js:75]: Just to clarify, the issue isn't that the button is used in a legacy or custom view. The issue is that the button is used in a view that can be rendered through a legacy widget (Usually the Form Dialog). 

In the case of res.partner, it's the res_partner_many2one widget that still hasn't been adapted. So any Studio form displaying a partner view could also be broken.

But the issue remain, I'm not sure how I'm supposed to find out which models can be called in that legacy FormView modal without checking every possible flow, or checking for each model that this button is on, if there's a field that reference them that is still rendered using the legacy view.

I'm sorry if this isn't completely clear, I have to admit that I haven't fully grasped how views are rendered yet.

If you have a suggestion on how I can find which module render this button in a legacy view, I wouldn't mind doing it though!

PS: I also asked the JS team on Discord if they had any advice on how to find which views are still rendered in legacy, and they told me that all views should be converted in 16.0 including studio if they're lucky. We could almost certainly remove this legacy button widget by then.

Comment (qsm-odoo) [reply addons/website/static/src/js/backend/button.js:75]: So if a legacy widget instantiate a non-legacy view which contains a website_redirect_button, it will use the legacy field? That seems... strange :grimacing: But I don't know more than you on the subject.

But ok, if you are convinved you cannot possibly know for sure where legacy `website_redirect_button` are still needed, we can move on without it and we'll see later... but then the plan is to wait for res_partner_many2one to be converted then we remove the legacy field and... cross fingers? :grimacing:

Comment (detrouxdev) [reply addons/website/static/src/js/backend/button.js:75]: > So if a legacy widget instantiate a non-legacy view which contains a website_redirect_button, it will use the legacy field? That seems... strange 😬 But I don't know more than you on the subject.

The legacy widget will instantiate a legacy view. Since the Partner view isn't a custom view, it renders it using the legacy FormView widget, and rendering the legacy buttons with it.

You can actually create the same issue with the website_publish_button that I removed. Its defined in the `gamification_badge_view_form` template. If I go in studio and make a many2one field that uses the res_partner_many2one widget  and link it to Gamification badge, the widget will not be loaded properly.
I just think it's so niche that it's OK to remove the legacy widget. Especially if the OWL team's goal is to converted all these widgets in 16.0.

Indeed, we'll have to cross fingers 🤞🏻  when we remove the button, or somehow find a way to make sure that we can no longer invoke those views via the legacy widget system.

File: addons/website/static/src/components/fields/redirect_field.xml [hunk 12]
```diff
@@ -1,16 +1,16 @@
...
+<t t-name="website.RedirectField" owl="1">
```
Comment (qsm-odoo) [inline addons/website/static/src/components/fields/redirect_field.xml:4]: @detrouxdev Except this RedictField template that was not touched by the other commit if I am not mistaken, the other reindentation should be squashed in the related commits where you basically rewrite them entirely anyway. The point of reindenting something in its own commit is to separe concerns in the history. But if we can avoid it... best do it (like taking the opportunity of a file being basically rewritten at 95% to reindent everything: do it there rather than on its own commit).

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
task-XXX mentioning the backend/frontend merge is wrong here: this has really nothing to do with it, right? Remove all task-XXX from your commits.

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Comment by qsm-odoo** [comment]
Forced a new batch
